src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -2527,6 +2527,152 @@ ITermDispatch::StringHandler AdaptDispatch::DownloadDRCS(const size_t fontNumber
+ITermDispatch::StringHandler AdaptDispatch::RequestSetting()
+{
+    // We use a VTIDBuilder to parse the characters in the control string into
+    // an ID which represents the setting being queried. If the given ID isn't
+    // supported, we respond with an error sequence: DCS 0 $ r ST. Note that
+    // this is the opposite of what is documented in most DEC manuals, which
+    // say that 0 is for a valid response, and 1 is for an error. The correct
+    // interpretation is documented in the DEC STD 070 reference.
+    const auto idBuilder = std::make_shared<VTIDBuilder>();
+    return [=](const auto ch) {
+        if (ch >= '\x40' && ch <= '\x7e')
+        {
+            const auto id = idBuilder->Finalize(ch);
+            switch (id)
+            {
+            case VTID('m'):
+                _ReportSGRSetting();
+                break;
+            case VTID('r'):
+                _ReportDECSTBMSetting();
+                break;
+            default:
+                _WriteResponse(L"\033P0$r\033\\");
+                break;
+            }
+            return false;
+        }
+        else
+        {
+            if (ch >= '\x20' && ch <= '\x2f')
+            {
+                idBuilder->AddIntermediate(ch);
+            }
+            return true;
+        }
+    };
+}
+
+// Method Description:
+// - Reports the current SGR attributes in response to a DECRQSS query.
+// Arguments:
+// - None
+// Return Value:
+// - None
+void AdaptDispatch::_ReportSGRSetting() const
+{
+    // A valid response always starts with DCS 1 $ r.
+    // Then the '0' parameter is to reset the SGR attributes to the defaults.
+    std::wstring response = L"\033P1$r0";
+
+    TextAttribute attr;
+    if (_pConApi->PrivateGetTextAttributes(attr))
+    {
+        // For each boolean attribute that is set, we add the appropriate
+        // parameter value to the response string.
+        const auto addAttribute = [&](const auto parameter, const auto enabled) {
+            if (enabled)
+            {
+                response += parameter;
+            }
+        };
+        addAttribute(L";1", attr.IsBold());
+        addAttribute(L";2", attr.IsFaint());
+        addAttribute(L";3", attr.IsItalic());
+        addAttribute(L";4", attr.IsUnderlined());
+        addAttribute(L";5", attr.IsBlinking());
+        addAttribute(L";7", attr.IsReverseVideo());
+        addAttribute(L";8", attr.IsInvisible());
+        addAttribute(L";9", attr.IsCrossedOut());
+        addAttribute(L";21", attr.IsDoublyUnderlined());
+        addAttribute(L";53", attr.IsOverlined());
+
+        // We also need to add the appropriate color encoding parameters for
+        // both the foreground and background colors.
+        const auto addColor = [&](const auto base, const auto color) {
+            const auto iterator = std::back_insert_iterator(response);
+            if (color.IsIndex16())
+            {
+                const auto index = XtermToWindowsIndex(color.GetIndex());
```

Comment:
Is this conversion backwards? Won't `GetIndex()` return the windows index, which we'll want to use as a VT index?

Reply:
Semantically yes, but in practice it's a bidirectional operation (if that's the right term). It's just swapping bits 0 and 2. I suppose we could create a couple of aliases for these functions to makes things more readable, but I really would like to just get rid of them one day.

Reply:
well I'll be danged, that does happen to work the other way, doesn't it